File: DSAlgorithms/Leetcode/c++/stacks/84-largest_rec_in_histogram.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        std::stack<std::vector<int>> a_stack;
        int max_area = 0;
        std::vector<int> top;
        std::vector<int> current_bar;
        for(int i = 0; i < heights.size(); i++)
        {
           int start = i;
           if(!a_stack.empty())
                top = a_stack.top();
           while(!a_stack.empty() && top[1] > heights[i]) 
           {
               top = a_stack.top();
               a_stack.pop();
               max_area = std::max(max_area, top[1] * (i - top[0]));
               start = top[0];
                if(!a_stack.empty())
                    top = a_stack.top();
           }
            current_bar.push_back(start);
            current_bar.push_back(heights[i]);
            a_stack.push(current_bar);
            current_bar.clear();
        }
        while(!a_stack.empty())
        {
            top = a_stack.top();
            a_stack.pop();
            int temp = top[1] * (heights.size() - top[0]);
            max_area = std::max(max_area, temp);
        }
        return max_area;
    }
};

File: DSAlgorithms/Leetcode/ts/src/stacks/84-largest_rec_in_histogram.ts
const largestRectangleArea = (heights: number[]): number =>{
    let max_area = 0 ;
    const a_stack: number[][] = [];
    let current_bar: number[] = [];
    let top: number[] = [];
    //console.log(a_stack[0][0])
    for(let i = 0; i < heights.length; i++)
    {
        let start = i;
        if(a_stack.length > 0)
            top = a_stack[a_stack.length - 1];
        while((a_stack.length > 0) && top[1] > heights[i])
        {
            top = a_stack[a_stack.length - 1];
            a_stack.pop();
            max_area = Math.max(max_area, top[1] * (i - top[0]))
            start = top[0];
            if(a_stack.length > 0)
                top = a_stack[a_stack.length - 1];
        }
        current_bar.push(start);
        current_bar.push(heights[i]);
        a_stack.push(current_bar);
        current_bar = [];
    }
    while(a_stack.length > 0)
    {
        top = a_stack[a_stack.length - 1];
        a_stack.pop();
        let temp = top[1] * (heights.length - top[0]);
        max_area = Math.max(max_area, temp);
    }
    return max_area;
};